tests/providers/hashicorp/secrets/test_vault.py

```diff
@@ -0,0 +1,137 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from unittest import TestCase, mock
+
+from hvac.exceptions import VaultError
+
+from airflow.providers.hashicorp.secrets.vault import VaultSecrets
+
+
+class TestVaultSecrets(TestCase):
+
+    @mock.patch("airflow.providers.hashicorp.secrets.vault.hvac")
+    def test_get_conn_uri(self, mock_hvac):
+        mock_client = mock.MagicMock()
+        mock_hvac.Client.return_value = mock_client
+        mock_client.secrets.kv.v2.read_secret_version.return_value = {
+            'request_id': '94011e25-f8dc-ec29-221b-1f9c1d9ad2ae',
+            'lease_id': '',
+            'renewable': False,
+            'lease_duration': 0,
+            'data': {
+                'data': {'test_postgres': 'postgresql://airflow:airflow@host:5432/airflow'},
```

Comment:
`test_postgres` here is in the duplicated in the key name isn't it? To make it easy to support storing in other forms (such as a k/v per attribute, i.e. not needing to URI parse) we could store that as 
```suggestion
                'data': {'conn_uri': 'postgresql://airflow:airflow@host:5432/airflow'},
```

WDYT?

Reply:
Right now this seems to be _all_ connections in a single path, with different values for each of them. I would have expected that each connection is it's own secret instead.

(I'm guessing Vault lets you limit access based on path, but not on key?)

Reply:
I.e. instead of this

```
vault kv put airflow/connections smtp_default=postgresql://airflow:airflow@host:5432/airflow
```

I'm proposing we do

```
vault kv put airflow/connections/smtp_default
conn_uri=postgresql://airflow:airflow@host:5432/airflow
```

Reply:
`conn_uri` looks redundant given we are already in `connections` path.

Reply:
Yes, that seems to be the level Vault polices operate at: https://www.vaultproject.io/docs/concepts/policies/#policy-syntax

Reply:
`conn_uri` is there because you need _something_ as the key, don't you?

Reply:
Current method:

```
vault kv put airflow/connections smtp_default=smtps://user:host@relay.example.com:465
vault kv put airflow/connections postgres_default=postgresql://airflow:airflow@host:5432/airflow
```

My proposal:

```
vault kv put airflow/connections/smtp_default conn_uri=smtps://user:host@relay.example.com:465
vault kv put airflow/connections conn_uri=postgresql://airflow:airflow@host:5432/airflow
```

And conn_uri is there so we could do this in the future:

```
vault kv put airflow/connections conn_type=postgresql user=airflow password=airflow host=host port=5432 schema=airflow
```

Reply:
Remember: I'm not familiar with Vault so may be making stupid assumptions.

Reply:
So all the connections would live in a single path `airflow/connections` and it accepts a key_value store.

Key I think should be **conn_id**.

Storing them as following:

```
vault kv put airflow/connections conn_type=postgresql user=airflow password=airflow host=host port=5432 schema=airflow
```

would create separate keys:

```
❯ vault kv put airflow/connections conn_type=mysql user=airflow password=airflow host=host port=5432 schema=airflow

Key              Value
---              -----
created_time     2020-03-18T22:00:51.665274Z
deletion_time    n/a
destroyed        false
version          2
❯ vault kv get airflow/connections
====== Metadata ======
Key              Value
---              -----
created_time     2020-03-18T22:00:51.665274Z
deletion_time    n/a
destroyed        false
version          2

====== Data ======
Key          Value
---          -----
conn_type    mysql
host         host
password     airflow
port         5432
schema       airflow
user         airflow
```

which doesn't make sense as there are not linked to each other

Reply:
By having them all in a single path does it restrict the ability to apply different policies to each key?

Reply:
I have worked with Google KMS and access was granted on a path 
Level too.

You generally grant a team permission to a path or mounting point, so they can add all rheir secrets in that path.

Reply:
Does G kms just have a single key at each path, or does it also have keys like valut does?

Reply:
Agree with what you have suggested on further reading and deep dive.

Updated: https://github.com/apache/airflow/pull/7741/commits/2311f937bcf14899314fe7bebf073da254f41d02